edgeApp/edgeSrc/NDPluginEdge.cpp

```diff
@@ -0,0 +1,366 @@
+    NDUInt8,    Unsigned 8-bit integer	
+    NDInt16,    Signed 16-bit integer	
+    NDUInt16,   Unsigned 16-bit integer	
+    NDInt32,    Signed 32-bit integer	
+    NDUInt32,   Unsigned 32-bit integer	
+    NDFloat32,  32-bit float		
+    NDFloat64   64-bit float		
+
+    NDColorModeMono,    Monochromatic image							
+    NDColorModeBayer,   Bayer pattern image, 1 value per pixel but with color filter on detector
+    NDColorModeRGB1,    RGB image with pixel color interleave, data array is [3, NX, NY]	
+    NDColorModeRGB2,    RGB image with row color interleave, data array is [NX, 3, NY]		
+    NDColorModeRGB3,    RGB image with plane color interleave, data array is [NX, NY, 3]	
+    NDColorModeYUV444,  YUV image, 3 bytes encodes 1 RGB pixel					
+    NDColorModeYUV422,  YUV image, 4 bytes encodes 2 RGB pixel					
+    NDColorModeYUV411   YUV image, 6 bytes encodes 4 RGB pixels					
+
+
+    NDArray         OpenCV
+    =========       ==========
+    NDInt8          CV_8S
+    NDUInt8         CV_8U
+    NDInt16         CV_16S
+    NDUInt16        CV_16U
+    NDInt32         CV_32S
+    NDUInt32        CV_32U
+    NDFloat32       CV_32F
+    NDFloat64       CV_64F
+
+    ND_BayerPatern  OpenCV
+    ==============  ======
+    NDBayer_RGGB    RG
+    NDBayer_GBRG    GB
+    NDBayer_GRGB    GR
+    NDBayer_BGGR    BG
+
+
+
+*/
+
+
+
+/** Callback function that is called by the NDArray driver with new NDArray data.
+  * Does image processing.
+  * \param[in] pArray  The NDArray from the callback.
+  */
+void NDPluginEdge::processCallbacks(NDArray *pArray)
+{
+    /* This function does array processing.
+     * It is called with the mutex already locked.  It unlocks it during long calculations when private
+     * structures don't need to be protected.
+     */
+    NDArray *pScratch=NULL;
+    NDArrayInfo arrayInfo;
+
+    int     i, j;
+    unsigned int     numRows, rowSize;
+    unsigned char *inData, *outData;
+    int    edge1;
+    int    edge1Found;
+    int    edge2;
+    int    edge2Found;
+
+    //static const char* functionName = "processCallbacks";
+
+
+    /* Call the base class method */
+    NDPluginDriver::processCallbacks(pArray);
+
+    /* make the array something we understand			*/
+    this->pNDArrayPool->convert( pArray, &pScratch, NDUInt8);
+
+    pScratch->getInfo(&arrayInfo);
+
+    rowSize = pScratch->dims[arrayInfo.xDim].size;
+    numRows = pScratch->dims[arrayInfo.yDim].size;
+
+    cv::Mat img = cv::Mat( numRows, rowSize, CV_8UC1);
+    // cv::Mat dst;
+    cv::Mat detected_edges;
+    double lowThreshold;
+    double thresholdRatio;
+    
+    getDoubleParam( NDPluginEdgeLowThreshold,   &lowThreshold);
+    getDoubleParam( NDPluginEdgeThresholdRatio, &thresholdRatio);
+
+    // We assume 8 bit mono image as input, at least for now.
+
+    inData  = (unsigned char *)pScratch->pData;
+    outData = (unsigned char *)img.data;
+    for( i=0; (unsigned int)i<arrayInfo.nElements; i++) {
```

Comment:
> My resident compiler expert tells me that modern compilers figure this sort
of thing out on their own and it is best to avoid memcpy.

Oh, I've never heard anyone suggesting to avoid memcpy before... It is true that modern compilers with optimization switched on will enable the program to automagically prefetch array data from RAM into the local CPU cache in chunks when detecting iterations over arrays; however memcpy is just as efficient in modern ages ASFAIK... We generally use memcpy around the areaDetector code-base whenever we need to copy blocks of data so I would go with that if for no other reason than sticking to convention :-)